### PR TITLE
feat: Add ActiveJob instrumentation

### PR DESCRIPTION
**Pull request by ahayworth**
This is a slightly modified version of the ActiveJob instrumentation
that we have been running for a few weeks now, on two different internal
applications. It is somewhat basic, but it works well enough.

There are a few things of note:
- We track a few custom attributes for ActiveJob - I'm not sure if the
  naming is good, however.
- ~We do not attempt to pass any kind of context propagation through the
  job system. I have successfully done that via job-argument-mangling,
  but didn't do so here. We've found that doing so leads to unwieldy
  traces for web requests that enqueue background jobs.~ edit: now we do, if you pass the `context_propagation: :child` argument. otherwise we use span links by default.
- I've tried to follow the overall messaging semantic conventions, since
  'job' semantic conventions have not materialized yet. I'll update this
  PR if they are finalized before this is merged; otherwise we can make
  those updates later.
- The test suite is a tad slow - to test a few features, I needed to
  explicitly use a real job system - in this case, the `async` adapter.
  It's all still in-process, but uses threads as a queueing system. I
  found the easiest thing to do was to just call `sleep` in a few places.

Probably the most important things to think about when reviewing are:
- Are the attributes added correct (or - are they at least close to
  correct until we have proper job semantic conventions)?
- ~Following a job produce -> job enqueue requires searching your tracing
  system on the value of `messaging.message_id`. Is that ergonomic or
  should we do something different?~ edit: span links are set by default, and users can opt-in to parent/child context semantics.

I'd appreciate anyone who's willing to test this out in a production
system - I think. We're going to test this version internally as well,
since it differs in a few small ways from what we actually run today.

<details>
<summary>Example span (async adapter)</summary>

```
❯ bundle exec ruby ./active_job.rb
I, [2021-04-23T15:45:02.363340 #43810]  INFO -- : Instrumentation: OpenTelemetry::Instrumentation::ActiveJob was successfully installed
#<struct OpenTelemetry::SDK::Trace::SpanData[ActiveJob] [TestJob] [ac093fd0-1221-4027-83f3-042643f7622b] Performing TestJob (Job ID: ac093fd0-1221-4027-83f3-042643f7622b) from Async(default) enqueued at 2021-04-23T20:45:02Z

 name="TestJob send",
 kind=:producer,
 status=
  #<OpenTelemetry::Trace::Status:0x00007fb35bc4a728 @code=1, @description="">,
 parent_span_id="\x00\x00\x00\x00\x00\x00\x00\x00",
 total_recorded_attributes=5,
 total_recorded_events=0,
 total_recorded_links=0,
 start_timestamp=2021-04-23 15:45:02.364016 -0500,
 end_timestamp=2021-04-23 15:45:02.364252 -0500,
 attributes=
  {"messaging.destination_kind"=>"queue",
   "messaging.system"=>"async",
   "messaging.destination"=>"default",
   "messaging.message_id"=>"ac093fd0-1221-4027-83f3-042643f7622b",
   "net.transport"=>"inproc"},
 links=nil,
 events=nil,
 resource=
  #<OpenTelemetry::SDK::Resources::Resource:0x00007fb35b9f2c00
   @attributes=
    {"service.name"=>"unknown_service",
     "process.pid"=>43810,
     "process.command"=>"./active_job.rb",
     "process.runtime.name"=>"ruby",
     "process.runtime.version"=>"2.7.2",
     "process.runtime.description"=>
      "ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin19]",
     "telemetry.sdk.name"=>"opentelemetry",
     "telemetry.sdk.language"=>"ruby",
     "telemetry.sdk.version"=>"0.17.0"}>,
 instrumentation_library=
  #<struct OpenTelemetry::SDK::InstrumentationLibrary
   name="OpenTelemetry::Instrumentation::ActiveJob",
   version="0.17.0">,
 span_id="\xAB\tY6\x04\x95r\xA1",
 trace_id="]W\x05\xD9UV`\xB3N\x93\xAF\xAAin~S",
 trace_flags=#<OpenTelemetry::Trace::TraceFlags:0x00007fb33b83e7f8 @flags=1>,
 tracestate=#<OpenTelemetry::Trace::Tracestate:0x00007fb35bbaa4d0 @hash={}>>
[ActiveJob] Enqueued TestJob (Job ID: ac093fd0-1221-4027-83f3-042643f7622b) to Async(default)

--------------------------------------------------
 The computer is doing some work, beep beep boop.
--------------------------------------------------

#<struct OpenTelemetry::SDK::Trace::SpanData
 name="TestJob process",
 kind=:consumer,
 status=
  #<OpenTelemetry::Trace::Status:0x00007fb35bc4a728 @code=1, @description="">,
 parent_span_id="\x00\x00\x00\x00\x00\x00\x00\x00",
 total_recorded_attributes=7,
 total_recorded_events=0,
 total_recorded_links=0,
 start_timestamp=2021-04-23 15:45:02.44189 -0500,
 end_timestamp=2021-04-23 15:45:02.441939 -0500,
 attributes=
  {"messaging.destination_kind"=>"queue",
   "messaging.system"=>"async",
   "messaging.destination"=>"default",
   "messaging.message_id"=>"ac093fd0-1221-4027-83f3-042643f7622b",
   "net.transport"=>"inproc",
   "messaging.operation"=>"process",
   "messaging.active_job.executions"=>1},
 links=nil,
 events=nil,
 resource=
  #<OpenTelemetry::SDK::Resources::Resource:0x00007fb35b9f2c00
   @attributes=
    {"service.name"=>"unknown_service",
     "process.pid"=>43810,
     "process.command"=>"./active_job.rb",
     "process.runtime.name"=>"ruby",
     "process.runtime.version"=>"2.7.2",
     "process.runtime.description"=>
      "ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin19]",
     "telemetry.sdk.name"=>"opentelemetry",
     "telemetry.sdk.language"=>"ruby",
     "telemetry.sdk.version"=>"0.17.0"}>,
 instrumentation_library=
  #<struct OpenTelemetry::SDK::InstrumentationLibrary
   name="OpenTelemetry::Instrumentation::ActiveJob",
   version="0.17.0">,
 span_id="\x9B:\x8A\xCC`\x91H\xED",
 trace_id="X\xE7\xE0\xCA\xA6\xA8\x1EG\x80J\x98\x91<!f\xAD",
 trace_flags=#<OpenTelemetry::Trace::TraceFlags:0x00007fb33b83e7f8 @flags=1>,
 tracestate=#<OpenTelemetry::Trace::Tracestate:0x00007fb35bbaa4d0 @hash={}>>
[ActiveJob] [TestJob] [ac093fd0-1221-4027-83f3-042643f7622b] Performed TestJob (Job ID: ac093fd0-1221-4027-83f3-042643f7622b) from Async(default) in 79.23ms
```

</details>

### CODE REVIEW

**Comment by ahayworth** [comment]
cc #222

**Comment by ahayworth** [comment]
Capturing notes from the meeting today:
- We're going to take another look at (optionally) passing context somehow, so that spans can (or cannot!) be linked into a parent/child or follows-from relationship.
- We will look into doing this in a way that could be generically useful to Rails and see if we think it makes sense to upstream that later on.
- I vaguely recall some kind of global request context feature in Rails that may be an interesting thing to use here.
- We will probably not do this via job argument mangling (if at all possible) since that can be error-prone.
- We are unsure if we will propagate the minimum amount of information for span relationship-building, or if we'll propagate the whole OpenTelemetry context.
- We'll revisit the sidekiq instrumentation (and possibly other instrumentation libraries) to understand this context somehow.

File: instrumentation/active_job/test/instrumentation/active_job/patches/active_job_callbacks_test.rb
Comment (arielvalentin) [inline instrumentation/active_job/test/instrumentation/active_job/patches/active_job_callbacks_test.rb:114]: It looks like you are using a mixture of the `Minitest::Assertions` along with `Minitest::Expectations`. I think for symmetry we should use `MiniTest::Expectations` here instead.

**Comment by arielvalentin** [comment]
@ahayworth ahh just saw this was in draft mode. Sorry for the premature feedback.

File: instrumentation/active_job/test/instrumentation/active_job/patches/active_job_callbacks_test.rb
Comment (ahayworth) [reply instrumentation/active_job/test/instrumentation/active_job/patches/active_job_callbacks_test.rb:51]: > 💭 I will do some research and see if we can test this without sleeps. I am concerned that this may result in non-deterministic tests.

That's certainly possible. Given that we specifically change behavior (to follow semantic conventions) to see if we're in a queue that executes in a separate thread (ie: not a test or inline job adapter), I don't have an easy answer here. We *could* just skip testing those bits, if we're concerned about the flaky tests or don't care all that much.

**Comment by ahayworth** [comment]
> @ahayworth ahh just saw this was in draft mode. Sorry for the premature feedback.

That's alright; I accidentally re-requested reviews when I tried to update my branch and committed a git foot-gun. 😄

**Comment by fbogsany** [comment]
A few general things:
1. I think we should try to propagate context, including baggage. *Not* doing that is quite limiting (e.g. for analysis involving both the enqueuing process and the job execution), unless ...
2. If ActiveJob has some kind of job ID, it could be used as a correlation ID, which would enable many analyses on *some* backends. Baggage has other benefits, though, as does trace context (e.g. for consistent sampling).
3. We should add a configuration option (to all job-system instrumentation, consistently named) to toggle between linking traces and continuing traces. I would tend to default to starting a new trace for the job execution with a link back to the trace for the enqueuing process. Links are not necessarily well supported by all trace backends, but they're part of the spec and should become more widely supported 🔜

**Comment by ahayworth** [comment]
Thanks for the feedback @fbogsany - I got 3/4 of the way through re-implementing it (and incidentally, doing what you suggested) but got distracted. This is a good reminder for me to finish. 😆 

> We should add a configuration option (to all job-system instrumentation, consistently named) to toggle between linking traces and continuing traces.

Do you have a feeling on what that name should be? And more specifically, is this something that should be supported per-job, or for all jobs traced via a given instrumentation library? The PR as it stands (which works, but tests are not fixed yet) allows for per-job configuration, but that may not be the right approach.

**Comment by fbogsany** [comment]
> Do you have a feeling on what that name should be?

I like my bike sheds yellow 😄. I don't have a strong opinion, or I won't until someone expresses one I don't like :trollface:.

> And more specifically, is this something that should be supported per-job, or for all jobs traced via a given instrumentation library? The PR as it stands (which works, but tests are not fixed yet) allows for per-job configuration, but that may not be the right approach.

At least a global option should be provided. I can sort of see the value of per-job configuration, but I am unlikely to reach for it.

**Comment by ahayworth** [comment]
That all makes sense. 🤣

File: instrumentation/active_job/test/instrumentation/active_job/patches/active_job_callbacks_test.rb
Comment (ahayworth) [reply instrumentation/active_job/test/instrumentation/active_job/patches/active_job_callbacks_test.rb]: Fixed in: 6a05a2f0
For some reason (and I cannot recall why now) I was getting deprecation warnings about `must_be_nil` - but I can't reproduce it ¯\\_(ツ)_/¯

File: instrumentation/active_job/test/instrumentation/active_job/patches/active_job_callbacks_test.rb
Comment (ahayworth) [reply instrumentation/active_job/test/instrumentation/active_job/patches/active_job_callbacks_test.rb:56]: Good point; I added it to `ensure` since restructuring it to do `after` blocks felt annoying. 😄

File: instrumentation/active_job/test/instrumentation/active_job/patches/active_job_callbacks_test.rb
Comment (ahayworth) [reply instrumentation/active_job/test/instrumentation/active_job/patches/active_job_callbacks_test.rb]: I fixed this up by shutting down the async executor thread in the tests where we switch to that adapter - there's no sleep required in that case. There's still _some_ chance for flakiness if the thread doesn't shut down correctly, but I think it's pretty drastically reduced.

I looked further into using the `:test` adapter, but it really expects to be running in a rails test suite. I started running into issues and realized I wasn't really saving any hassle at that point.

**Comment by ahayworth** [comment]
@open-telemetry/ruby-approvers this is ready for review (I noticed a redis test failure, but all activejob tests are passing). I believe I have a reasonable solution to serializing context along with the job, by re-using the built-in argument serializing capabilities of ActiveJob. Unless I've misunderstood things (and I may have), I believe this gets baggage too @fbogsany - please let me know if that's not the case.

I chose a global option for how to configure propagation, and it defaults to actually just linking to the enqueueing trace rather than setting up a parent/child relationship. One weird note is that the `:inline` queue adapter renders the context propagation moot; and the job execution is always in the same trace where it was enqueued. That's because ActiveJob executes the job while in the process of enqueueing it when using the `:inline` adapter; and honestly I didn't feel like jumping through a ton of hoops to detect that case and do something about it. In any case, if you're using the `:inline` adapter then I can't imagine why you wouldn't want your job's execution to be in the same logical trace, anyways.

I've been testing an older version of this internally, and it seemed okay - but I'll deploy this to production and make sure nothing unexpected blows up - but I think it's ready now!

**Comment by ahayworth** [comment]
I'm not certain if the span names I've chosen actually follow the messaging semantic attributes: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#span-name

I think having the job class as part of the span name makes it a lot more useful, since there's no other way in the semantic conventions to define the job name; and the background-job semantic conventions PR [has not yet been approved](https://github.com/open-telemetry/opentelemetry-specification/pull/1582). I'm curious what people think we should do:

- We could keep what we have. In backend systems that display the span name, this works nicely (it's not so helpful in other systems...)
- We could adhere more closely to the messaging conventions and add the class name as something like `active_job.job: FooJob` in the span attributes, which may be the "best of both worlds approach"
- We could venture out on a limb and apply the not-yet-merged Job semantics, if we think they're likely to be accepted.

Curious what others think here.

**Comment by fbogsany** [comment]
In the Sidekiq instrumentation, we added an option to use the job class name as the span name: https://github.com/open-telemetry/opentelemetry-ruby/blob/bda1fb057c63f174c81c687631a35559362c7e26/instrumentation/sidekiq/lib/opentelemetry/instrumentation/sidekiq/instrumentation.rb#L30
I think that's a reasonable compromise.

**Comment by ahayworth** [comment]
@fbogsany I agree, and added that option in 0cfe562. I'm fighting with the test suite (yay) and the cop (yay) but I think the PR is otherwise in a decent state.

**Comment by ahayworth** [comment]
Fun:
```
opentelemetry-instrumentation-active_job: test
C:/hostedtoolcache/windows/Ruby/3.0.1/x64/lib/ruby/3.0.0/logger/log_device.rb:103:in `initialize': No such file or directory @ rb_sysopen - /dev/null (Errno::ENOENT)
```

File: instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb
Comment (robertlaurin) [inline instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb:48]: This is an interesting addition I want to think about a little.  I have it a backlog todo list to add shutdown hooks to sidekiq as part of the instrumentation so every application owner isn't responsible for adding at as part of its config.  

I wonder if we should do the same for resque instead of pushing the responsibility of it onto a place like this.

Comment (robertlaurin) [reply instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb:48]: Barring my above comment I'm wonder if we should consider this type of scenario `absolutely necessary`.  

>ForceFlush SHOULD only be called in cases where it is absolutely necessary, such as when using some FaaS providers that may suspend the process after an invocation, but before the SpanProcessor exports the completed spans.

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#forceflush-1

File: instrumentation/active_job/.ruby-version
Comment (ahayworth) [reply instrumentation/active_job/.ruby-version]: 🤦 I tried so hard not to accidentally commit that. *I* need it, but people who don't need to run 5 versions of ruby in different codebases every day probably do not! 😆 

Removed in 7984c23c !

File: instrumentation/active_job/test/test_helper.rb
Comment (ahayworth) [reply instrumentation/active_job/test/test_helper.rb:15]: Ah good call.

File: instrumentation/active_job/test/test_helper.rb
Comment (ahayworth) [reply instrumentation/active_job/test/test_helper.rb]: If you don't see a big problem with it, I am fine leaving it in as well!

File: instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb
Comment (ahayworth) [reply instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb:81]: I had trouble re-installing the instrumentation after changing config options in test. But ... really the truth is that I'm actually just bad at writing tests. I see how you did it in the resque PR and I'll adapt that method! 😄

Comment (ahayworth) [reply instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb:81]: I updated things in aca3d053 based on reading the resque instrumentation tests. I've found that the test suite seems to fail if I re-install over and over again, which I suspect is because ActiveJob doesn't like getting `patch`ed over and over again (suddenly tests which expect 3 spans start seeing ... a lot of spans). I don't have solid evidence for that, though. Lemme know what you think about the approach in aca3d053 and if that's any better/worse than what came before. 😆

File: instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/instrumentation.rb
Comment (ahayworth) [reply instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/instrumentation.rb]: Ah cool, removed!

File: instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb
Comment (ahayworth) [reply instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb:48]: > Barring my above comment I'm wonder if we should consider this type of scenario absolutely necessary.

Yes, actually I would say this would qualify as `absolutely necessary`. If you don't flush, you run the very real chance of resque (or another forking job processor) killing things before the export deadline rolls around for a `BatchSpanProcessor`. Maybe even worth updating the spec to mention it.

In a really ideal world we could add an exit hook for `resque` to flush rather than doing it after a job is processed in ActiveJob, but we'd also want to do that for other queues as well (sidekiq does this kind of thing too, I think?). Maybe carrying patches for just those two would be okay in this instrumentation, but I would start to feel rather odd doing it for more than just one or two queueing systems. Feels like a lot of work... but in a bad way - more like "not sure if that's going to cause us headaches or not later one".

I don't think we could really rely on the sidekiq or resque instrumentation adding necessary shutdown hooks either; there's no real guarantee that someone loaded those instrumentations.

If we decide that this is not a good idea, we could possibly make it optional - either an instrumentation option, or maybe check for an env variable or something.

File: instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb
Comment (robertlaurin) [reply instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb]: That makes sense, I think the approach taken is reasonable one given multiple patches don't behave nicely.

File: instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb
Comment (robertlaurin) [inline instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb:44]: I think the instrumentation patch here will probably need the same treatment as the resque PR did https://github.com/open-telemetry/opentelemetry-ruby/pull/779#discussion_r645895665

The TLDR is we lose baggage propagation when using `:link` or `:none`

Comment (ahayworth) [reply instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb]: This should be working now.

**Comment by ahayworth** [comment]
Note that after #820 was merged, in 982e7a6 we now need to call `OpenTelemetry.tracer_provider.force_flush`, rather than `ActiveJob::Instrumentation.instance.tracer.tracer_provider.force_flush`. I think that's still safe, but I'm not sure.

File: instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb
Comment (fbogsany) [inline instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb:60]: I still have doubts about this as the right approach, but not enough to block merging the PR. Would you consider adding a configuration option for this?

Comment (ahayworth) [reply instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb]: Implemented in 35039f80

**Review by fbogsany — APPROVED** [review]
Thanks @ahayworth - this looks awesome!

**Pull request merged**
